File: src/intent_pred_pkg/scripts/intent_prediction_node.py
#!/usr/bin/env python3
import rospy
from visualization_msgs.msg import MarkerArray
from std_msgs.msg import Int32  # Add this import
import numpy as np
import tf2_ros
import tf2_geometry_msgs
from geometry_msgs.msg import TransformStamped, PointStamped

from intent_prediction_calc import IntentPredictCalc

class IntentPredictor:
    def __init__(self):
        rospy.init_node('intent_predictor')

        # Initializing TF2 Buffer and Listener
        self.tf_buffer = tf2_ros.Buffer(rospy.Duration(10.0))
        self.tf_listener = tf2_ros.TransformListener(self.tf_buffer)

        # Defining frame names
        self.hand_depth_frame = "hand_camera_depth_optical_frame"
        self.head_depth_frame = "head_camera_depth_optical_frame"

        # initialize the intent predict algorithm class
        self.intent_predict_calc = IntentPredictCalc()
        
        # Subscribe to the centroid markers from PCLProcessor
        self.centroid_sub = rospy.Subscriber('/centroid_markers', MarkerArray, self.centroid_callback)
        
        # Publish the selected object ID
        self.intent_pub = rospy.Publisher('/object_id_to_grasp', Int32, queue_size=1)

        # Waiting for transforms to become available
        rospy.loginfo("Waiting for transforms to become available...")
        self._wait_for_transforms()

        # Add debug timer to periodically report transform status
        rospy.Timer(rospy.Duration(10.0), self._debug_transforms)
        
    def centroid_callback(self, marker_array):
        num_markers = len(marker_array.markers)

        if num_markers == 0:
            rospy.logwarn("Received empty marker array")
            return
        
        # get transform from hand depth frame to head depth frame
        try:
            transform = self.tf_buffer.lookup_transform(
                self.head_depth_frame, # target frame
                self.hand_depth_frame, # source frame
                rospy.Time(0), # get latest available transform
                rospy.Duration(1.0) # wait up to 1 second for transform
            )
            rospy.logdebug_throttle(5.0, "Got transform from hand to head depth frame")
        except (tf2_ros.LookupException, tf2_ros.ConnectivityException, tf2_ros.ExtrapolationException) as e:
            rospy.logwarn(f"Failed to lookup transform: {e}")
            return
        
        # Pre-allocate arrays
        centroids = np.zeros((num_markers, 3), dtype=np.float32)
        labels = np.zeros(num_markers, dtype=np.int32)
        transformed_centroids = np.zeros((num_markers, 3), dtype=np.float32)  # New array for transformed points
        
        # Fill arrays (normal + transformed centroids)
        for i, marker in enumerate(marker_array.markers):
            # rriginal position in hand camera frame
            position = marker.pose.position
            centroids[i] = [position.x, position.y, position.z]
            labels[i] = marker.id

            # transform the centroid to the head camera depth frame
            point = PointStamped()
            point.header.frame_id = self.hand_depth_frame
            point.point.x = position.x
            point.point.y = position.y
            point.point.z = position.z

            transformed_point = tf2_geometry_msgs.do_transform_point(point, transform)

            transformed_centroids[i] = [
                transformed_point.point.x,
                transformed_point.point.y,
                transformed_point.point.z
            ]
        
        # Run the intent prediction algorithm
        intended_idx = self.intent_predict_calc.find_grasp_obj(centroids, transformed_centroids, labels)

        # check if the index is valid
        if 0 <= intended_idx < len(labels):
            self.intent_pub.publish(int(labels[intended_idx]))
            rospy.loginfo(f"Predicted grasp object ID: {labels[intended_idx]}")
        else:
            rospy.logwarn(f"Invalid prediction index: {intended_idx}")

    def _wait_for_transforms(self):
        """Wait for the required transforms to be available"""
        frames_ready = False
        while not frames_ready and not rospy.is_shutdown():
            try:
                # Check if we can transform between the frames
                self.tf_buffer.lookup_transform(
                    self.head_depth_frame,
                    self.hand_depth_frame,
                    rospy.Time(0),
                    rospy.Duration(1.0)
                )
                frames_ready = True
                rospy.loginfo("All required transforms are available.")
            except (tf2_ros.LookupException, tf2_ros.ConnectivityException, tf2_ros.ExtrapolationException):
                rospy.logwarn_throttle(5.0, "Waiting for transforms between camera frames...")
                rospy.sleep(0.5)

    def _debug_transforms(self, event=None):
        """Print debug info about the TF tree"""
        try:
            transform = self.tf_buffer.lookup_transform(
                self.head_depth_frame,
                self.hand_depth_frame,
                rospy.Time(0),
                rospy.Duration(0.1)
            )
            t = transform.transform.translation
            r = transform.transform.rotation
            rospy.loginfo(f"Transform hand→head: [{t.x:.3f}, {t.y:.3f}, {t.z:.3f}], quat=[{r.x:.3f}, {r.y:.3f}, {r.z:.3f}, {r.w:.3f}]")
        except Exception as e:
            rospy.logwarn(f"Debug transform check failed: {e}")

if __name__ == '__main__':
    predictor = IntentPredictor()
    rospy.spin()

File: src/intent_pred_pkg/src/intent_prediction_node.cpp
#include "intent_pred_pkg/intent_prediction_node.h"

IntentPredictor::IntentPredictor() 
    : nh_("~"),
      tf_listener_(tf_buffer_)
{
    // Get parameters (with defaults)
    nh_.param<std::string>("hand_depth_frame", hand_depth_frame_, "cam1_depth_optical_frame");
    nh_.param<std::string>("head_depth_frame", head_depth_frame_, "cam2_depth_optical_frame");
    
    // Initialize the intent prediction algorithm
    intent_predict_calc_ = std::make_unique<IntentPredictCalc>();
    
    ROS_INFO("Waiting for transforms to become available...");
    if (!waitForTransforms()) {
        ROS_ERROR("Failed to get required transforms. Exiting.");
        ros::shutdown();
        return;
    }
    
    // Initialize subscribers
    centroid_sub_ = nh_.subscribe("/centroid_markers", 10, 
        &IntentPredictor::centroidCallback, this);
    
    // Initialize publishers
    intent_pub_ = nh_.advertise<std_msgs::Int32>("/object_id_to_grasp", 10);
    
    // Initialize debug timer (10 second period)
    debug_timer_ = nh_.createTimer(ros::Duration(10.0), 
        &IntentPredictor::debugTransformsCallback, this);
    
    ROS_INFO("Intent predictor node initialized successfully");
}

bool IntentPredictor::waitForTransforms()
{
    // Poll until transform is available or timeout
    const int max_attempts = 20;
    const float poll_interval = 0.5;
    
    for (int attempt = 0; attempt < max_attempts; ++attempt) {
        try {
            tf_buffer_.lookupTransform(
                head_depth_frame_,
                hand_depth_frame_,
                ros::Time(0),
                ros::Duration(1.0)
            );
            ROS_INFO("All required transforms are available.");
            return true;
        } catch (const tf2::TransformException& ex) {
            ROS_WARN_STREAM_THROTTLE(5.0, "Waiting for transforms: " << ex.what());
            ros::Duration(poll_interval).sleep();
        }
    }
    
    ROS_ERROR("Timed out waiting for transforms.");
    return false;
}

void IntentPredictor::debugTransformsCallback(const ros::TimerEvent& event)
{
    try {
        geometry_msgs::TransformStamped transform = tf_buffer_.lookupTransform(
            head_depth_frame_,
            hand_depth_frame_,
            ros::Time(0),
            ros::Duration(0.1)
        );
        
        const auto& t = transform.transform.translation;
        const auto& r = transform.transform.rotation;
        
        ROS_INFO_STREAM("Transform hand→head: [" 
            << t.x << ", " << t.y << ", " << t.z << "], quat=["
            << r.x << ", " << r.y << ", " << r.z << ", " << r.w << "]");
            
    } catch (const tf2::TransformException& ex) {
        ROS_WARN_STREAM("Debug transform check failed: " << ex.what());
    }
}

void IntentPredictor::centroidCallback(const visualization_msgs::MarkerArray::ConstPtr& marker_array)
{
    const size_t num_markers = marker_array->markers.size();
    
    if (num_markers == 0) {
        ROS_WARN("Received empty marker array");
        return;
    }
    
    // Get transform from hand depth frame to head depth frame
    geometry_msgs::TransformStamped transform;
    try {
        transform = tf_buffer_.lookupTransform(
            head_depth_frame_,  // target frame
            hand_depth_frame_,  // source frame
            ros::Time(0),       // latest available transform
            ros::Duration(1.0)  // wait up to 1 second
        );
    } catch (const tf2::TransformException& ex) {
        ROS_WARN_STREAM("Failed to lookup transform: " << ex.what());
        return;
    }
    
    // Pre-allocate data structures
    std::vector<Eigen::Vector3f> centroids(num_markers);
    std::vector<Eigen::Vector3f> transformed_centroids(num_markers);
    std::vector<int> labels(num_markers);
    
    // Process each marker
    for (size_t i = 0; i < num_markers; ++i) {
        const auto& marker = marker_array->markers[i];
        
        // Original position in hand camera frame
        centroids[i] = Eigen::Vector3f(
            marker.pose.position.x,
            marker.pose.position.y,
            marker.pose.position.z
        );
        
        labels[i] = marker.id;
        
        // Transform point to head camera frame
        geometry_msgs::PointStamped point_in, point_out;
        point_in.header.frame_id = hand_depth_frame_;
        point_in.point.x = marker.pose.position.x;
        point_in.point.y = marker.pose.position.y;
        point_in.point.z = marker.pose.position.z;
        
        tf2::doTransform(point_in, point_out, transform);
        
        transformed_centroids[i] = Eigen::Vector3f(
            point_out.point.x,
            point_out.point.y,
            point_out.point.z
        );
    }
    
    // Run the intent prediction algorithm
    int intended_idx = intent_predict_calc_->findGraspObj(centroids, transformed_centroids, labels);
    
    // Check if the index is valid
    if (intended_idx >= 0 && intended_idx < static_cast<int>(labels.size())) {
        std_msgs::Int32 msg;
        msg.data = labels[intended_idx];
        intent_pub_.publish(msg);
        ROS_INFO_STREAM("Predicted grasp object ID: " << labels[intended_idx]);
    } else {
        ROS_WARN_STREAM("Invalid prediction index: " << intended_idx);
    }
}

int main(int argc, char** argv)
{
    ros::init(argc, argv, "intent_prediction_node");
    IntentPredictor predictor;
    ros::spin();
    return 0;
}